src/connection.h

```diff
@@ -245,14 +359,74 @@ int connDisableTcpNoDelay(connection *conn);
 int connKeepAlive(connection *conn, int interval);
 int connSendTimeout(connection *conn, long long ms);
 int connRecvTimeout(connection *conn, long long ms);
-int connPeerToString(connection *conn, char *ip, size_t ip_len, int *port);
-int connFormatFdAddr(connection *conn, char *buf, size_t buf_len, int fd_to_str_type);
-int connSockName(connection *conn, char *ip, size_t ip_len, int *port);
-const char *connGetInfo(connection *conn, char *buf, size_t buf_len);
 
 /* Helpers for tls special considerations */
-sds connTLSGetPeerCert(connection *conn);
-int tlsHasPendingData();
-int tlsProcessPendingData();
+void *connTypeGetCtx(const char *typename);
+void *connTypeGetClientCtx(const char *typename);
+
+/* Get cert for the secure connection */
+static inline sds connGetPeerCert(connection *conn) {
+    if (conn->type->get_peer_cert) {
+        return conn->type->get_peer_cert(conn);
+    }
+
+    return NULL;
+}
+
+/* Initialize the redis connection framework */
+int connTypeInitialize();
+
+/* Register a connection type into redis connection framework */
+int connTypeRegister(ConnectionType *ct);
+
+/* Lookup a connection type by type name */
+ConnectionType *connectionByType(const char *typename);
+
+/* Fast path to get TCP connection type */
+ConnectionType *connectionTypeTcp();
+
+/* Lookup the index of a connection type by type name, return -1 if not found */
+int connectionIndexByType(const char *typename);
+
+/* Create a connection of specified type */
+connection *connCreate(const char *typename);
+
+/* Create a accepted connection of specified type.
+ * @priv is connection type specified argument */
+connection *connCreateAccepted(const char *typename, int fd, void *priv);
+
+/* Configure a connection type. A typical case is to configure TLS.
+ * @priv is connection type specified,
+ * @reconfigure is boolean type to specify if overwrite the original config */
+int connTypeConfigure(const char *typename, void *priv, int reconfigure);
+
+/* Cleanup a connection type. A typical case is to cleanup TLS. */
+void connTypeCleanup(const char *typename);
```

Comment:
i still feel that there are still way too many interfaces that take a string as a connection type, and i think that all of these should take a `ConnectionType*` or index as input.
i feel that there should be only two interfaces that take that string, one is the registration (connCreate) and the other is a lookup (connectionByType) that should be **only** used at startup to cache direct pointers to some types (or by using a singleton method like you did with  connectionTypeTcp)

Reply:
OK!
By the way, `void connTypeCleanup(const char *typename);` is unused now, it's supposed to clean up a single listener, but we clean up all the listeners by `void connTypeCleanupAll()`. So I'll remove this function in the next version.
Also introduce connectionTypeTls() to get TLS connection type fastly.

Reply:
i still see interfaces that i think i'd like to change / delete.
* connCreateAccepted (should take a connection type)
* connTypeOfCluster (should return a connection type)
* connCreate (should take a connection type)
* connTypeOfReplication (should return a connection type)
* connTestType - delete, if someone needs to use it, it may be an indication that they're doing the wrong thing?
* connAcceptHandler - same?